### DISCUSSION TITLE
GDScript → Add conditional compilation directives

### DISCUSSION THREAD

**Discussion by azur-wolve** [godotengine/godot-proposals#4023] [Scripting]
C has `#if` that you can use to perform conditional compilation.
It is useful to make a separation between "testing mode" and "normal mode".
The core idea is to have some lines of codes (checks, print statements and so on) that would be ignored when exporting the project by disabling it → `#TESTING_MODE = false`
Using another syntax compatible with GDScript.

Another way could be use a keyword to "mark" blocks of code.
Could be by creating some "special" `bool` variable and then use it to "mark" blocks of code.

Example:
```gdscript
# in the header section where the other fields are declared..
<keyword> conditional_var = true

# somewhere in the code
conditional_var:
# a block of nested code
```

The difference with a normal `bool` would be that it tells the compiler to ignore it if false. Useful to release/export and so on.

It would be declared mostly inside **Autoloads**.

(Probably something like that had been already proposed.)

**Comment by Kazetsukai** [comment]
One method that works quite well already to achieve this kind of thing is features. You can set different features per export, and check whether those are set.

![image](https://user-images.githubusercontent.com/935494/154795042-2219f9c3-b862-451b-a7f8-5abb39dfa537.png)

There is even a default feature set that tells you if you are running in debug mode or not, which can be very useful.

![image](https://user-images.githubusercontent.com/935494/154795083-058aa154-bfe5-481f-b914-37384d896b80.png)

**Reply by azur-wolve** [reply]
when using

```gdscript
if OS.has_feature("debug"):
    print("text")
```
It would not compile/ship to the exported binary file?

**Reply by Kazetsukai** [reply]
It will be in the bytecode. Is there a reason for wanting it not to be?

**Reply by azur-wolve** [reply]
Not a particular one. Not pretty sure on how convenient is to have many `if` statements out there, aren't they adding overhead?

**Reply by lentsius-bark** [reply]
Depending on the complexity of the checks I imagine the overhead may at times be unwanted and improve performance

**Comment by Mickeon** [comment]
For the sake of completion, I'm going to mention the proposal for [conditional compilation with Feature Tags](https://github.com/godotengine/godot-proposals/issues/4234) here, as well.

**Comment by setanarut** [comment]
Useful but bad idea that reduces code readability. Makes it difficult to read code written by others. I love the philosophy that less keyword is better. Go language should be taken as an example

**Comment by dalexeev** [comment]
I made a plugin: [dalexeev/gdscript-preprocessor](https://github.com/dalexeev/gdscript-preprocessor).
